server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/aws/AwsS3ObjectStorage.java

```diff
@@ -170,54 +178,84 @@ private File createTempFile(Blob blob) throws IOException {
+            return puttingAttempt
                 .publishOn(Schedulers.elastic())
                 .retryWhen(Retry
-                    .<Void>onlyIf(retryContext -> needToCreateBucket(retryContext.exception()))
+                    .onlyIf(retryContext -> failOnNoSuchBucket(retryContext.exception(), bucketName))
                     .exponentialBackoff(FIRST_BACK_OFF, FOREVER)
                     .withBackoffScheduler(Schedulers.elastic())
                     .retryMax(MAX_RETRY_ON_EXCEPTION)
-                    .doOnRetry(retryContext -> createBucket(bucketName, configuration)));
+                    .doOnRetry(retryContext -> createBucket(bucketName, configuration)))
+                .retryWhen(Retry
+                    .onlyIf(retryContext -> failOnExistingBucket(retryContext.exception(), bucketName))
+                    .withBackoffScheduler(Schedulers.elastic())
+                    .exponentialBackoff(FIRST_BACK_OFF, FOREVER)
+                    .retryMax(RETRY_ONE_LAST_TIME_ON_CONCURRENT_SAVING));
         }
 
-        private void uploadByFile(ObjectStorageBucketName bucketName, BlobId blobId, File file) throws InterruptedException {
-            PutObjectRequest request = new PutObjectRequest(bucketName.asString(), blobId.asString(), file);
-            upload(request);
+        private Mono<Void> uploadByFile(ObjectStorageBucketName bucketName, BlobId blobId, File file) {
+            return Mono.fromSupplier(() -> new PutObjectRequest(bucketName.asString(), blobId.asString(), file))
+                .flatMap(this::upload);
         }
 
-        private void uploadByBlob(ObjectStorageBucketName bucketName, Blob blob) throws InterruptedException, IOException {
-            try (InputStream payload = blob.getPayload().openStream()) {
-                PutObjectRequest request = new PutObjectRequest(bucketName.asString(),
-                    blob.getMetadata().getName(),
-                    payload,
-                    new ObjectMetadata());
+        private Mono<Void> uploadByBlob(ObjectStorageBucketName bucketName, Blob blob) {
+            return Mono.fromCallable(() -> blob.getPayload().openStream())
+                .flatMap(inputStream -> upload(new PutObjectRequest(
+                        bucketName.asString(),
+                        blob.getMetadata().getName(),
+                        inputStream,
+                        new ObjectMetadata()))
+                    .doFinally(ignored -> closeQuietly(inputStream)));
+        }
 
-                upload(request);
+        private Mono<Void> upload(PutObjectRequest request) {
+            return Mono.create(sink -> {
+                TransferManager transferManager = getTransferManager();
+                transferManager
+                    .upload(request)
+                    .addProgressListener((ProgressListener) progressEvent -> watchingUploadEvent(sink, transferManager, progressEvent));
```

Comment:
why is the point using that API?

Reply:
It's a version used to compare with the fully async implemented in the next commit. Asked by Benoit here https://github.com/linagora/james-project/pull/3023#discussion_r361319808